---
title: "Population parameters and selection scan"
author: "Gen Xu"
date: "02-12-2025"
output:
  prettydoc::html_pretty:
    theme: architect
    highlight: github
---
## Sequence alignment and prepare for the bam files
- Index the Reference Genome
```{bash,eval=F}
bwa index reference.fa
```
- Mapping reads to the reference genome
```{bash,eval=F}
bwa mem -t 16 reference.fa sample_R1.fastq.gz sample_R2.fastq.gz > sample.sam
```
- Converting SAM to BAM
```{bash,eval=F}
module load samtools
samtools view -bS sample.sam > sample.bam
samtools sort sample.bam -o sample.sorted.bam
samtools index sample.sorted.bam
```
- Removing Duplicates Using Picard
```{bash,eval=F}
module load picard/2.18
picard -Xmx60g MarkDuplicates I=sample.sorted.bam O=sample.sorted.picard.bam M=sample.sorted.picard.matrics AS=true 
```
## Here is the [**link**](https://app.globus.org/file-manager?origin_id=02e69d09-39f8-44f8-be56-ef476410cc89&origin_path=%2F) for the demo data
## [**ANGSD**](https://www.popgen.dk/angsd/index.php/ANGSD): Analysis of next generation Sequencing Data
#### Please go to computation node
```{bash,eval=F}
srun --qos=short --nodes=1 --licenses=common --ntasks=4 --mem 32G --licenses=common --time 6:00:00 --pty bash
```

### Calculating nucleotide diversity
- Step1: Index Your Reference Genome
```{bash,eval=F}
module load samtools angsd
cd 01bam_files/
samtools faidx maize-genome-V5.fa
```
- Step 2: Prepare BAM File List

- Step 3: Generate Site Frequency Spectrum (SFS)
```{bash,eval=F}
angsd -bam bam_list_mazie.txt  -doMaf 1 -doMajorMinor 1 -uniqueOnly 1 -minMapQ 30 -minQ 20 -minInd 6 -doSaf 1 -anc maize-genome-V5.fa -GL 2 -out maize -P 4

realSFS maize.saf.idx -fold 1 -P 4 > maize.sfs
```
- Step 4: Compute π
Now, use the estimated SFS to calculate nucleotide diversity:
```{bash,eval=F}
realSFS saf2theta maize.saf.idx -outname maize -sfs maize.sfs -fold 1 -P 4
```

- Step 5: Extract π estimates using a 25 kb window size with a 5 kb step size.
```{bash,eval=F}
thetaStat do_stat maize.thetas.idx -win 25000 -step 5000 -outnames maize.thetasWindow25kb
```
** In the pestPG file are the sum of the per site estimates for a region
** If you don't have the ancestral states, you can still calculate the Watterson and Tajima theta, which means you can perform the Tajima's D neutrality test statistic. But this requires you to use the folded sfs. The output files will have the same format, but only the thetaW and thetaD, and tajimas D is meaningful.
** We then have 5 different estimators of theta, these are: Watterson, pairwise, FuLi, fayH, L. And we have 5 different neutrality test statistics: Tajima's D, Fu&Li F's, Fu&Li's D, Fay's H, Zeng's E. The final column is the effetive number of sites with data in the window.
###plot results
```{r,eval=F}
library(data.table)
d=fread("maize.thetasWindow25kb.pestPG",header=T,data.table=F)[,-1]
d=d[d[,1]==10,]
pi=d$tP/d$nSites
plot(d$tW/d$nSites,d$tP/d$nSites,xlab="Watterson Theta",ylab="Pairwise Theta",pch=16,col=adjustcolor("blue",alpha.f = 0.5))
plot(d$WinCenter/1e6,d$tP/d$nSites,xlab = "Chr10 (Mb)",ylab="Pairwise Theta")
```

### Single read sampling approach for PCA---Corrects for unequal sequencing depth bias
```{bash,eval=F}
angsd -bam bam_list_mazie.txt -minMapQ 30 -minQ 20 -GL 2  -doMajorMinor 1 -doMaf 1 -SNP_pval 2e-6 -doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 -out maize -P 4
```
- plot PCA
```{r}
name <- "maize.covMat"
m <- as.matrix(read.table(name))
eig <- eigen(m)
write.table(eig$vectors, "eigenvectors.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(eig$values, "eigenvalues.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
plot(eig$vectors[,1:2],lwd=2,xlab="PC 1",ylab="PC 2",main="Principal components",col="darkgreen",pch=16,las=1,cex=2)
```

## Powerful tools for population parameter calculation: [**Plink**](https://www.cog-genomics.org/plink/1.9/) and [**vcftools**](https://vcftools.github.io/man_latest.html)
### Functions in plink
- Prepare binary files for plink
```{bash,eval=F}
cd ../02vcf_files/
gunzip -c TEO_LR_MZ_test.vcf.gz > TEO_LR_MZ_test.vcf
module load plink
plink --vcf TEO_LR_MZ_test.vcf --make-bed --out  TEO_LR_MZ_test
```
- Minor allele frequency
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --freq --out TEO_LR_MZ_test
```
- Missing data
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --missing --out TEO_LR_MZ_test
```
- Hardy-Weinberg equilibrium
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --hardy --out TEO_LR_MZ_test
```
- Homozygous genotype counts for each sample
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --het --out TEO_LR_MZ_test
```
- multiple parameters
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --freq --missing --hardy --het --out TEO_LR_MZ_test
```

- PCA
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --pca 10 --out TEO_LR_MZ_test
```
- Variant pruning
```{bash,eval=F}
plink --bfile TEO_LR_MZ_test --indep-pairwise 100 10 0.1 --out outfile
plink --bfile TEO_LR_MZ_test --extract outfile.prune.in --make-bed --out prunedData
```
- bfiles to vcf
```{bash,eval=F}
plink --bfile prunedData --recode vcf --out prunedData
```
### Functions in Vcftools
- Allele frequency from Chr5
```{bash,eval=F}
vcftools --gzvcf TEO_LR_MZ_test.vcf.gz --freq --chr 10 --out chr10_analysis
```
- Filtering according SNP missing rate or sample name
```{bash,eval=F}
vcftools --vcf TEO_LR_MZ_test.vcf --max-missing 0.7 --recode --out TEO_LR_MZ_test_miss03

vcftools --vcf TEO_LR_MZ_test.vcf --keep maize.txt --recode --out maize
```

- Nucleotide diversity
```{bash,eval=F}
vcftools --vcf maize.recode.vcf --window-pi 25000 --window-pi-step 5000 --out MZ_test
```
- Fst estimate from Weir and Cockerham’s 1984 paper
```{bash,eval=F}
vcftools --vcf TEO_LR_MZ_test.vcf --weir-fst-pop teo.txt --weir-fst-pop maize.txt --fst-window-size 25000 --fst-window-step 5000 --out TEO_vs_MZ_w25k
```
- Small sample size and low allele frequency can casue negative fst, we can set them to 0
```{bash,eval=F}
awk '{if($5 < 0) $5=0; print}' TEO_vs_MZ_w25k.windowed.weir.fst > TEO_vs_MZ_w25k.windowed_fixed.fst
```

### Selection scan using [**XP-CLR**](https://github.com/hardingnj/xpclr)
The original paper is [**here**](https://pubmed.ncbi.nlm.nih.gov/20086244/)
```{bash,eval=F}
module load xpclr/1.1
xpclr --out /work/jyanglab/gxu6/12-19-2024/teaching932_selection/test_files/02vcf_files/TEO_MZ_Chr10 --input TEO_LR_MZ_test.vcf --format vcf --samplesA maize.txt --samplesB teo.txt --chr 10 --ld 0.7 --start 98000000 --stop 105000000 --maxsnps 200 --minsnps 200 --size 25000 --step 5000
```
- Object is pop A. Reference is pop B
- --out full path is needed