---
title: "Angsd and XP-CLR"
author: "Jinliang Yang"
date: "2025-03-13"
output:
  html_document:
    df_print: paged
  word_document: default
---


# A procedure to calculate $\theta$ values using ANGSD


## Here is the [**link**](https://app.globus.org/file-manager?origin_id=02e69d09-39f8-44f8-be56-ef476410cc89&origin_path=%2F) for the demo data


## [**ANGSD**](https://www.popgen.dk/angsd/index.php/ANGSD): Analysis of next generation Sequencing Data

#### Request a computation node

```{bash,eval=F}
srun --qos=short --nodes=1 --licenses=common --ntasks=1 --mem 6G --licenses=common --time 2:00:00 --pty bash
```

### Calculating nucleotide diversity

- Step1: Index Your Reference Genome

```{bash,eval=F}
module load samtools angsd
cd largedata/sorghum/sorted_bam
samtools faidx chr1.fasta
```

- Step 2: Prepare BAM File List

```{bash,eval=F}
ls sorted*bam > bam_list.txt
```

- Step 3: Generate Site Frequency Spectrum (SFS)

```{bash,eval=F}
angsd -bam bam_list.txt  -doMaf 1 -doMajorMinor 1 -uniqueOnly 1 -minMapQ 30 -minQ 20 -minInd 6 -doSaf 1 -anc chr1.fasta -GL 2 -out sorghum -P 4

realSFS sorghum.saf.idx -fold 1 -P 4 > sorghum.sfs
```

- Step 4: Compute theta

Now, use the estimated SFS to calculate nucleotide diversity:
```{bash,eval=F}
realSFS saf2theta sorghum.saf.idx -outname sorghum -sfs sorghum.sfs -fold 1 -P 4
```

- Step 5: Extract theta estimates using a 5 kb window size with a 1 kb step size.

```{bash,eval=F}
thetaStat do_stat sorghum.thetas.idx -win 5000 -step 1000 -outnames sorghum.thetasWindow5kb_1kb

cp sorghum.thetasWindow5kb_1kb.pestPG ../../../cache/
```

- In the pestPG file are the sum of the per site estimates for a region
- If you don't have the ancestral states, you can still calculate the Watterson and Tajima theta, which means you can perform the Tajima's D neutrality test statistic. But this requires you to use the folded sfs. The output files will have the same format, but only the thetaW and thetaD, and tajimas D is meaningful.

- We then have 5 different estimators of theta, these are: 
  - Watterson, pairwise, FuLi, fayH, L. 
  
- And we have 5 different neutrality test statistics: Tajima's D, Fu&Li F's, Fu&Li's D, Fay's H, Zeng's E. 

- The final column is the effective number of sites with data in the window.

### plot results

```{r,eval=F}
library(data.table)
d <- fread("cache/sorghum.thetasWindow5kb_1kb.pestPG", header=T, data.table=F)

pi=d$tP/d$nSites
plot(d$tW/d$nSites,d$tP/d$nSites,xlab="Watterson Theta",ylab="Pairwise Theta",pch=16,col=adjustcolor("blue",alpha.f = 0.5))
plot(d$WinCenter/1000, d$tP/d$nSites, xlab = "Chr1 (kp)", ylab="Pairwise Theta")
```

```{r}
plot(d$WinCenter/1000, d$Tajima/d$nSites, xlab = "Chr1 (kp)", ylab="Tajima's D")
```

### an example about how to interpret the results in text

#### 1. First, provide a brief description of the analysis, but include details such as data, software, and parameters in the 'Materials and Methods' section.

In this simulated experiment, we generated 10 individuals from each population (see M&M for details).

#### 2. Begin by describing the results, and if necessary, provide a brief reminder or explanation to help educate the reader.

Tajima's D, a measure of the site frequency spectrum (SFS) distribution between two populations, is expected to be around 0 if a genomic region is neutral. As shown in the figure, most of the sites have values close to 0, consistent with our simulation, which did not include any sites under selection.

#### 3. Highlight any exciting discoveries.

However, we observed that in the 10 kb region, the values of Tajima's D are slightly above 0, with an average of approximately 0.03. A positive Tajima's D can indicate balancing selection, as it suggests higher-than-expected nucleotide diversity. This typically occurs because both alleles have been maintained at the balanced sites in the population, allowing more mutations to accumulate. 

#### 4. Offer alternative explanations, can be backed up by other's studies

However, this slight elevation above 0 is more likely caused by a high mutation rate in the simulation, consistent with previous studies (citation).



