---
title: "Compute Fst using sorghum and its genomic features"
author: "Jinliang Yang"
date: "2025-02-26"
output:
  html_document:
    df_print: paged
  word_document: default
---


# A procedure to calculate $\theta$ values

### 2. Obtain the SNP calls  with `samtools`

```{bash}
#!/bin/bash -l
#SBATCH -D /mnt/nrdstor/jyanglab/jyang21/2025-agro932-lab
#SBATCH -o /mnt/nrdstor/jyanglab/jyang21/2025-agro932-lab/slurm-log/stdout-%A_%a.txt
#SBATCH -e /mnt/nrdstor/jyanglab/jyang21/2025-agro932-lab/slurm-log/stderr-%A_%a.txt
#SBATCH -J SNP_calling
#SBATCH -t 4:00:00
#SBATCH --array=1-20  # Define an array job from 1 to 20
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1  # Adjust based on available resources

set -e
set -u

# Load required modules
module load samtools bcftools

# Change to the working directory
cd largedata/sorghum

# Define sample ID from the SLURM array index
SAMPLE_ID=$SLURM_ARRAY_TASK_ID

# Step 1: Index the sorted BAM file (if not already indexed)
if [ ! -f sorted_l${SAMPLE_ID}.bam.bai ]; then
    samtools index sorted_l${SAMPLE_ID}.bam
fi

# Step 2: Generate mpileup file
bcftools mpileup -f chr1.fasta sorted_l${SAMPLE_ID}.bam > l${SAMPLE_ID}.mpileup

# Step 3: Call SNPs with conserved calling and output everything
bcftools call -c -Ob -o l${SAMPLE_ID}.bcf l${SAMPLE_ID}.mpileup

# Step 4: Index the VCF file
bcftools index l${SAMPLE_ID}.bcf

```

submit job using the below code:

```{bash, eval=FALSE}
sbatch --licenses=common slurm-script/snpcalling.sh
```


# Merging All VCF Files (After All Jobs Complete)

```{bash}
module load bcftools
bcftools merge -Ob -o merged.bcf l{1..20}.bcf
bcftools index merged.bcf

# get a summary of the VCF file
bcftools stats merged.bcf > bcf_summary.txt

# Filter for Biallelic SNPs
bcftools view merged.bcf -m2 -M2 -v snps -Ob -o merged_biallelic_snps.bcf


# get a summary again
bcftools stats  merged_biallelic_snps.bcf > bcf_sum2.txt
```


# Convert VCF to Tab-Delimited Format

```{bash}
bcftools query -f '%CHROM\t%POS\t%REF\t%ALT\t%QUAL\t%DP[\t%GT]\n' merged_biallelic_snps.bcf > snp_calls.txt
```
Chromosome (CHROM)
Position (POS)
Reference allele (REF)
Alternate allele (ALT)
Quality (QUAL)
Depth of coverage (DP)
Genotypes (GT) for each sample

# Convert VCF to PLINK format

```{bash}
module load plink/1.90

plink --bcf merged_biallelic_snps.bcf --allow-extra-chr --set-missing-var-ids @:# --make-bed --out merged_variants

# move merged files to a new folder
mkdir merged
mv merged_varaints* merged/
cd merged/

# get the missing data and maf PCA
plink --bfile merged_variants --allow-extra-chr --missing --freq --pca 20 --out merged_variants

```





